examples/api/lib/widgets/text_magnifier/text_magnifier.0.dart

```diff
@@ -0,0 +1,105 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+
+void main() => runApp(const MyApp());
+
+class MyApp extends StatelessWidget {
+  const MyApp({super.key});
+
+  static const String textFieldText = 'Custom Magnifier!';
+  static const Size loupeSize = Size(200, 200);
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      home: Scaffold(
+        body: Padding(
+          padding: const EdgeInsets.symmetric(horizontal: 48.0),
+          child: Center(
+            child: TextField(
+              // Create a custom magnifier configuration that
+              // this `TextField` will use to build a magnifier with.
+              magnifierConfiguration: TextMagnifierConfiguration(
+                magnifierBuilder: (_, __, ValueNotifier<MagnifierInfo> magnifierInfo) => CustomMagnifier(
+                  magnifierInfo: magnifierInfo,
+                ),
+              ),
+              controller: TextEditingController(text: textFieldText),
+            ),
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+class CustomMagnifier extends StatelessWidget {
+  const CustomMagnifier({super.key, required this.magnifierInfo});
+
+  static const Size magnifierSize = Size(200, 200);
+
+  // This magnifier will consume some text data and position itself
+  // based on the info in the magnifier.
+  final ValueNotifier<MagnifierInfo> magnifierInfo;
+
+  @override
+  Widget build(BuildContext context) {
+    // Use a value listenable builder because we want to rebuild
+    // every time the text selection info changes.
+    // `CustomMagnifier` could also be a `StatefulWidget` and call `setState`
+    // when `magnifierInfo` updates. This would be useful for more complex
+    // positioning cases.
+    return ValueListenableBuilder<MagnifierInfo>(
+        valueListenable: magnifierInfo,
+        builder: (BuildContext context,
+            MagnifierInfo currentMagnifierInfo, _) {
+          // We want to position the magnifier at the global position of the gesture.
+          Offset magnifierPosition = currentMagnifierInfo.globalGesturePosition;
+
+          // You may use the `MagnifierInfo` however you'd like:
+          // In this case, we make sure the magnifier never goes out of the current line bounds.
+          magnifierPosition = Offset(
+            clampDouble(
```

Comment:
I think this will crash if the text is RTL? since `left` will be greater than `right` (they probably should be called start/end or leading/trailing)

Reply:
(`clampDouble` asserts that `min <= max`)

Reply:
Current line boundaries does not care about LTR / RTL; https://github.com/antholeole/flutter/blob/29f51194a57bc0f049d424127107bb14791247b4/packages/flutter/lib/src/widgets/selectable_region.dart#L587-L588 It just uses the size of the text field. I will write a quick test for this, though, since I had the same thought (and many others down the line might too)

Reply:
`TextField` doesn't use selectable region. The implementation here seems to use the leading/trailing and the rect it produces may have a negative width:

https://github.com/flutter/flutter/blob/0906c14dd725a457ec5b5b0540665be6ff899561/packages/flutter/lib/src/widgets/text_selection.dart#L555-L575

Reply:
I vaguely recall Justin and I talking about this during the initial implementation, but I could be incorrect: I believe that this is still LTR/RTL agnostic, because `baseOffset` should be the right in RTL and the left in LTR, and visa versa for extent offset. The implementation of [`Rect.fromPoints`](https://github.com/antholeole/flutter/blob/27dc2c77ff0979043152c8afdb145fe0c254b037/bin/cache/pkg/sky_engine/lib/ui/geometry.dart#L681-L686) doesn't care which corners you pass in. So what happens in RTL is we pass in the top right and bottom left points, and in LTR we pass in top left and bottom right. 

Lots of RTLs and LTRs! either way I wrote a test using directionality RTL (see latest commit) and it seems to work okay.

Reply:
Ah you're right `Rect.fromPoints` guarantees the Rect doesn't have negative dimensions. `baseOffset` and `extentOffset` don't know about the writing direction. `extentOffset` is the edge that moves if you press an arrow key on the keyboard: https://master-api.flutter.dev/flutter/services/TextSelection-class.html

Reply:
ohh right. but to get the `TextSelection` we are using to calculate line boundaries, we do `renderEditable.getLineAtOffset` which should (if I'm reading the code right) select the whole line, so it only matters that base and extent point to the opposite ends of the line.